r1cs/gadgets/std/src/bits/boolean.rs

```diff
@@ -808,7 +892,16 @@ impl<ConstraintF: Field> EqGadget<ConstraintF> for Boolean {
             // 1 - 1 = 0 - 0 = 0
             (Constant(true), Constant(true)) | (Constant(false), Constant(false)) => return Ok(()),
             // false != true
-            (Constant(_), Constant(_)) => return Err(SynthesisError::AssignmentMissing),
+            (Constant(_), Constant(_)) => {
+                if should_enforce.is_constant() {
```

Comment:
What is the utility of this if?
Why is it not present in conditional_enforce_not_equal.

Reply:
It was a bugfix for a bug I found out when running unit tests I wrote for `UIntGadget`. In the previous implementation, if `self` and `other` were constants with different values, `conditional_enforce_equal` would simply return an error, independently  from the value of `should_enforce` flag, which was wrong. Therefore, I modified the code to handle all the possible cases depending on the `should_enforce` variable.

The bug was triggered in my unit tests only for `conditional_enforce_equal` function, but I forgot to check if the same bug exists also for `conditional_enforce_not_equal`, which explains why the code differs between the 2 functions. Thanks to your comment, I found out that the same bug should exist also for `conditional_enforce_not_equal`, so I will modify the code accordingly

Reply:
While reviewing the code of `conditional_enforce_not_equal` to fix the bug for the case when `self` and `other` are both constants with the same value, I found out that the constraint enforced at the end of the function was wrong. Therefore I had to modify the constraint, which also required to add the `eval_lc` function the `ConstraintSystemAbstract` trait to compute the value of a linear combination of variables in the constraint system. I have also modified the unit test for `conditional_enforce_equal` to test also the `conditional_enforce_not_equal` function, which was not tested before, explaining why we missed the incorrect constraint previously enforced in such function